feat(ui): path prefix support via HTTP header

mgoltzsche opened the pull request:
**Description**

Makes the web app honour the `X-Forwarded-Prefix` HTTP request header that may be sent by a reverse-proxy in order to inform the app that its public routes contain a path prefix. This allows to serve the webapp via a reverse-proxy/ingress controller under a path prefix/sub path such as e.g. `/localai/` while still being able to use the regular LocalAI routes/paths without prefix when directly connecting to the LocalAI server.

Changes:
* Add new `StripPathPrefix` middleware to strip the path prefix (provided with the `X-Forwarded-Prefix` HTTP request header) from the request path prior to matching the HTTP route.
* Add a `BaseURL` utility function to build the base URL, honouring the `X-Forwarded-Prefix` HTTP request header.
* Generate the base URL using the `BaseURL` function into the HTML (`head.html` and `login.html` templates) as `<base/>` tag.
* Make all webapp-internal URLs (within HTML+JS) relative in order to make the browser resolve them against the `<base/>` URL specified within each HTML page's header.
* Make font URLs within the CSS files relative to the CSS file.
* Generate redirect location URLs using the new `BaseURL` function.
* Use the new `BaseURL` function to generate absolute URLs within gallery JSON responses.

This PR fixes #3095

TL;DR:
The header-based approach allows to move the path prefix configuration concern completely to the reverse-proxy/ingress as opposed to having to align the path prefix configuration between LocalAI, the reverse-proxy and potentially other internal LocalAI clients.
The gofiber swagger handler already supports path prefixes this way, see [here](https://github.com/gofiber/swagger/blob/e2d9e9916d8809e8b23c4365f8acfbbd8a71c4cd/swagger.go#L79).

Known limitations:
* gofiber swagger does not support executing API requests when a path prefix is used because it doesn't replace the base URL within the served swagger JSON, see https://github.com/gofiber/swagger/issues/104.
* gofiber swagger does not support using a web app with multiple path prefixes simultaneously, see https://github.com/gofiber/swagger/issues/105.

**Notes for Reviewers**
Apart from the added automated tests, I did some manual testing and grepping, hoping to have all URLs covered.

**[Signed commits](../CONTRIBUTING.md#signing-off-on-commits-developer-certificate-of-origin)**
- [x] Yes, I signed my commits.
 
<!--
Thank you for contributing to LocalAI! 

Contributing Conventions
-------------------------

The draft above helps to give a quick overview of your PR.

Remember to remove this comment and to at least:

1. Include descriptive PR titles with [<component-name>] prepended. We use [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/).
2. Build and test your changes before submitting a PR (`make build`). 
3. Sign your commits
4. **Tag maintainer:** for a quicker response, tag the relevant maintainer (see below).
5. **X/Twitter handle:** we announce bigger features on X/Twitter. If your PR gets announced, and you'd like a mention, we'll gladly shout you out!

By following the community's contribution conventions upfront, the review process will 
be accelerated and your PR merged more quickly.

If no one reviews your PR within a few days, please @-mention @mudler.
-->

mudler commented:
Thank you :100:  - it makes totally sense as an addition to me. As soon as I have a bit of time will try to test out locally too to see if there aren't any regression

mgoltzsche commented:
The app test was broken. I think I fixed it now. Please re-run the workflow.

mgoltzsche commented:
Oups, turns out I introduced a cyclic dependency within the test code between the new `utils` package and the `middleware` package when I also added the `<base/>` URL to the login page that is rendered by the auth middleware.
I'll fix that...

mgoltzsche commented:
I fixed it by simplifying the `BaseURL` implementation and making its unit test not depend on the `StripPathPrefix` middleware. Please re-run the workflow.
